go-controller/pkg/ovn/egressgw.go

```diff
@@ -0,0 +1,255 @@
+		if err != nil {
+			return fmt.Errorf("unable to unmarshall annotation k8s.v1.cni.cncf.io/network-status on pod %s: %v", pod.Name, err)
+		}
+		for _, multusNetwork := range multusNetworks {
+			if multusNetwork.Name == pod.Annotations[routingNetworkAnnotation] {
+				for _, gwIP := range multusNetwork.IPs {
+					ip := net.ParseIP(gwIP)
+					if ip != nil {
+						foundGws = append(foundGws, ip)
+					}
+				}
+			}
+		}
+	} else if pod.Spec.HostNetwork {
+		for _, podIP := range pod.Status.PodIPs {
+			ip := net.ParseIP(podIP.IP)
+			if ip != nil {
+				foundGws = append(foundGws, ip)
+			}
+		}
+	} else {
+		klog.Errorf("Ignoring pod %s as an external gateway candidate. Invalid combination "+
+			"of host network: %t and routing-network annotation: %s", pod.Name, pod.Spec.HostNetwork, routingNetworkAnnotation)
+		return nil
+	}
+
+	// if we found any gateways then we need to update current pods routing in the relevant namespace
+	if len(foundGws) == 0 {
+		klog.Warningf("No valid gateway IPs found for requested external gateway pod: %s", pod.Name)
+		return nil
+	}
+
+	for _, namespace := range strings.Split(routingNamespaceAnnotation, ",") {
+		err := oc.addPodExternalGWForNamespace(namespace, pod, foundGws)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addPodExternalGWForNamespace handles adding routes to all pods in that namespace for a pod GW
+func (oc *Controller) addPodExternalGWForNamespace(namespace string, pod *kapi.Pod, gws []net.IP) error {
+	nsInfo, err := oc.waitForNamespaceLocked(namespace)
+	if err != nil {
+		return err
+	}
+	defer nsInfo.Unlock()
+	nsInfo.routingExternalPodGWs[pod.Name] = gws
+	return oc.addGWRoutesForNamespace(namespace, gws, nsInfo)
+}
+
+// addExternalGWsForNamespace handles adding annotated gw routes to all pods in namespace
+// This should only be called with a lock on nsInfo
+func (oc *Controller) addExternalGWsForNamespace(gateways []net.IP, nsInfo *namespaceInfo, namespace string) error {
+	if gateways == nil {
+		return fmt.Errorf("unable to add gateways routes for namespace: %s, gateways are nil", namespace)
+	}
+	nsInfo.routingExternalGWs = gateways
+	return oc.addGWRoutesForNamespace(namespace, gateways, nsInfo)
+}
+
+// addGWRoutesForNamespace handles adding routes for all existing pods in namespace
+// This should only be called with a lock on nsInfo
+func (oc *Controller) addGWRoutesForNamespace(namespace string, gws []net.IP, nsInfo *namespaceInfo) error {
+	existingPods, err := oc.watchFactory.GetPods(namespace)
+	if err != nil {
+		return fmt.Errorf("failed to get all the pods (%v)", err)
+	}
+	for _, pod := range existingPods {
+		gr := "GR_" + pod.Spec.NodeName
+		for _, gw := range gws {
+			for _, podIP := range pod.Status.PodIPs {
+				mask := GetIPFullMask(podIP.IP)
+				_, stderr, err := util.RunOVNNbctl("--", "--may-exist", "--policy=src-ip", "--ecmp-symmetric-reply",
```

Comment:
This should have been there from the beginning, but OVN upstream packages didn't have support yet. I can break it into a second commit to make it more obvious. --ecmp-symmetric-reply basically ensures that return ingress traffic from a pod goes back to the originating gateway (instead of being load balanced to another gateway):
https://bugzilla.redhat.com/show_bug.cgi?id=1849683